File: scripts/causal-neutral-driver.test.py
"""Deterministic tools qualification; real Node loads stubs only, never archived runRow."""
import copy
import importlib.util
import json
import os
from pathlib import Path
import subprocess
import tempfile
import unittest
from unittest.mock import patch

HERE=Path(__file__).resolve().parent

def module(name,path):
    spec=importlib.util.spec_from_file_location(name,path)
    value=importlib.util.module_from_spec(spec)
    spec.loader.exec_module(value)
    return value

runner=module('crossover_runner',HERE/'causal-neutral-driver.py')
verifier=module('crossover_verifier',HERE/'verify-causal-neutral-driver.py')
MUTANTS=[]


def samples():
    result=[]
    now=0
    for batch in range(3):
        for arm in (['reference','candidate'] if batch==1 else ['candidate','reference']):
            for index in range(400):
                duration=99999 if index<100 else (index-99)*(batch+1)*(2 if arm=='candidate' else 1)
                result.append(dict(batch=batch,arm=arm,index=index,phase='warmup' if index<100 else 'measured',
                    segment=f'cold-P2-summary/{batch}/{arm}/{index}',start=now,end=now+duration,
                    ms=duration,constructionMs=0,preparationMs=0))
                now+=duration+1
    return result


def write(path,value):
    path.write_text(json.dumps(value)+'\n')


def job_fixture(root,cell='H0',position=0):
    job=root/f'{position}-{cell}'; job.mkdir()
    kind,reference,order=runner.CELLS[cell]
    write(job/'config.json',runner.config_for(job,cell))
    (job/'entry.mjs').write_text(runner.entry_source(job,cell))
    events=[dict(kind='loaded',module=m,pid=123) for m in ['N',0,1]]+[dict(kind='driver',module='N' if kind=='neutral' else reference,pid=123)]
    (job/'entry-events.jsonl').write_text(''.join(json.dumps(x)+'\n' for x in events))
    write(job/'exit.json',dict(pid=123,exitCode=0,cwd=str(job),argv=['node',*runner.FLAGS,str(job/'entry.mjs')],startMonotonic=10,endMonotonic=11))
    write(job/'worker.json',dict(pid=123,node='v24.18.0',control=True,row=runner.ROW,
        recipe=dict(warmup=100,measured=300,orders=[['candidate','reference'],['reference','candidate'],['candidate','reference']])))
    write(job/'preflight.json',dict(passed=True))
    if kind=='neutral':
        write(job/'preflight-first.json',dict(module=order[0],pid=123))
        write(job/'preflight-order.json',dict(cell=cell,referenceModule=reference,preflightOrder=order,pid=123))
    write(job/'completion.json',dict(completed=True,samples=2400))
    (job/'stdout.log').write_text('PRESET_PERFORMANCE_ROW_DONE cold-P2-summary 2400\n')
    (job/'samples.jsonl').write_text(''.join(json.dumps(s)+'\n' for s in samples()))
    return job


def index_fixture(root):
    write(root/'artifact-index.json',dict(files={str(p.relative_to(root)):runner.digest(p.read_bytes())
        for p in root.rglob('*') if p.is_file() and p!=root/'artifact-index.json'}))


def full_fixture(root):
    materials,bindings=runner.archive_materials()
    for name,data in materials.items():(root/name).write_bytes(data)
    (root/'source').mkdir()
    source_paths=runner.SOURCE_PATHS
    for p in source_paths:(root/'source'/p.name).write_bytes(p.read_bytes())
    (root/'approval.json').write_bytes(runner.APPROVAL.read_bytes())
    subprocess.run(['node',str(HERE/'derive-causal-neutral-driver.mjs'),str(root/'worker.mjs'),str(root/'neutral-driver.mjs'),str(root/'derivation.json')],check=True,capture_output=True,timeout=15)
    result=subprocess.run(['node',str(HERE/'check-causal-neutral-source.mjs'),str(root/'worker.mjs'),str(root/'neutral-driver.mjs')],check=True,capture_output=True,text=True,timeout=15)
    write(root/'source-check.json',json.loads(result.stdout))
    write(root/'reservation.json' ,dict(schedule=runner.SCHEDULE,runtime=dict(node='v24.18.0',v8='13.6.233.17-node.50'),
        implicitNodeOptions={},sourceDigests={p.name:runner.digest(p.read_bytes()) for p in source_paths},
        approvalDigest=runner.digest(runner.APPROVAL.read_bytes()),executionRoot=str(root),executable='node'))
    rows=[]
    for position,cell in enumerate(runner.SCHEDULE):
        job=job_fixture(root,cell,position)
        exit_record=runner.read(job/'exit.json')
        exit_record.update(position=position,cell=cell,startMonotonic=10+position*2,endMonotonic=11+position*2)
        write(job/'exit.json',exit_record)
        rows.append(dict(position=position,cell=cell,**runner.summarize(samples(),cell)))
    write(root/'result.json',dict(status='complete-diagnostic',stopReason=None,notRun=[],elapsedSeconds=24,
        rows=rows,performanceQualification=False,eventAssociation='unknown: retained native clocks uncalibrated'))
    index_fixture(root)


class NeutralTests(unittest.TestCase):
    def test_all_coordinates_and_quantiles(self):
        for cell in runner.CELLS:
            with tempfile.TemporaryDirectory() as tmp:
                root=Path(tmp).resolve();job=job_fixture(root,cell)
                got=verifier.verify_job(job,cell,root)
                self.assertEqual(got,runner.summarize(samples(),cell))
                self.assertEqual(got['medianBatchP95Ms'],dict(candidate=1140,reference=570))
                self.assertEqual(got['candidateOverReference'],2)
                self.assertEqual(got['m1OverM0'],2 if cell.endswith('0') else .5)
                self.assertEqual(got['preflightLaterOverEarlier'],.5 if cell.startswith('R') else 2)
                self.assertEqual(got['hostedExternalOverLocal'],2 if cell.startswith('H') else None)

    def test_00_all_retained_real_cold_samples_before_synthetic_acceptance(self):
        import tarfile
        totals={}
        for name in ['causal-performance-repetition-v2','causal-control-role-crossover-attempt-2','causal-control-role-crossover-v1']:
            archive=HERE.parent/'archive/evals'/name/'evidence.tar.gz'
            total=0
            with tarfile.open(archive) as tar:
                for member in tar.getmembers():
                    if member.name.endswith('/samples.jsonl'):
                        rows=[json.loads(line) for line in tar.extractfile(member).read().decode().splitlines()]
                        if rows and rows[0]['segment'].startswith('cold-P2-summary/'):
                            expected=[]
                            for batch,arms in enumerate([['candidate','reference'],['reference','candidate'],['candidate','reference']]):
                                for arm in arms+(['plain'] if len(rows)==3600 else []):
                                    for index in range(400):expected.append((batch,arm,index,'warmup' if index<100 else 'measured',f'cold-P2-summary/{batch}/{arm}/{index}'))
                            self.assertEqual([(s['batch'],s['arm'],s['index'],s['phase'],s['segment']) for s in rows],expected)
                            for sample in rows:
                                self.assertEqual(sample['constructionMs'],0)
                                self.assertEqual(sample['preparationMs'],0)
                                self.assertEqual(sample['ms'],sample['end']-sample['start'])
                            paired=[s for s in rows if s['arm']!='plain']
                            verifier.v.verify_samples(paired,0)
                            runner.h.summarize(paired,0);total+=len(rows)
            totals[name]=total
        self.assertEqual(totals,{'causal-performance-repetition-v2':67200,'causal-control-role-crossover-attempt-2':19200,'causal-control-role-crossover-v1':2400})

    def test_actual_entries_and_pre_dispatch_rejection(self):
        for cell in runner.CELLS:
            for mutation in [None,'row','control','slot','cell','order','self-copy','driver']:
                with tempfile.TemporaryDirectory() as tmp:
                    root=Path(tmp).resolve();job=root/'job';job.mkdir();actual=root/'actual.jsonl'
                    common='import {appendFileSync} from "node:fs";\n'+f'const mark=x=>appendFileSync({json.dumps(str(actual))},JSON.stringify(x)+"\\n");\n'
                    for id,name in enumerate(runner.NAMES):
                        (root/name).write_text(common+f'mark(["import",{id}]);\nexport const id={id};\nexport function preflight(){{mark(["preflight",id]);}}\nexport function measurementArm(){{mark(["factory",id]);return {{cleanup(){{mark(["cleanup",id]);}}}};}}\n'+
                            'export async function runRow(p){ const {readFileSync}=await import("node:fs"); const c=JSON.parse(readFileSync(p)); const twin=await import(c.copyModule); preflight(); twin.preflight(); measurementArm().cleanup(); twin.measurementArm().cleanup(); mark(["driver",id]); }')
                    (root/'neutral-driver.mjs').write_text(common+'mark(["import","N"]);\nexport async function runRow(p,modules){ const {readFileSync}=await import("node:fs"); const c=JSON.parse(readFileSync(p)); for(const i of c.diagnostic.preflightOrder)modules[i].preflight(); for(const m of modules)m.measurementArm().cleanup(); mark(["driver","N"]); }')
                    config=runner.config_for(job,cell);entry=runner.entry_source(job,cell)
                    if mutation=='row':config['row']={**config['row'],'group':'recovery'}
                    if mutation=='control':config['control']=False
                    if mutation=='slot':config['diagnostic']['referenceModule']=1-config['diagnostic']['referenceModule']
                    if mutation=='cell':config['diagnostic']['cell']='invalid'
                    if mutation=='order':config['diagnostic']['preflightOrder']=list(reversed(config['diagnostic']['preflightOrder']))
                    if mutation=='self-copy':config['copyModule']=(root/runner.NAMES[runner.CELLS[cell][1]]).as_uri()
                    if mutation=='driver':entry=entry.replace('await n.runRow','await m0.runRow') if not cell.startswith('H') else entry.replace('await m'+cell[-1]+'.runRow','await n.runRow')
                    write(job/'config.json',config);(job/'entry.mjs').write_text(entry)
                    child=subprocess.run(['node',str(job/'entry.mjs')],capture_output=True,text=True,timeout=10,env={'PATH':os.environ['PATH']})
                    observed=[json.loads(line) for line in actual.read_text().splitlines()] if actual.exists() else []
                    if mutation is None:
                        self.assertEqual(child.returncode,0,child.stderr)
                        self.assertEqual(observed[:3],[['import','N'],['import',0],['import',1]])
                        self.assertEqual([x[1] for x in observed if x[0]=='preflight'],runner.CELLS[cell][2])
                        self.assertEqual(sorted(x[1] for x in observed if x[0]=='factory'),[0,1])
                        self.assertEqual(sorted(x[1] for x in observed if x[0]=='cleanup'),[0,1])
                        self.assertEqual(observed[-1],['driver',int(cell[-1]) if cell.startswith('H') else 'N'])
                    elif mutation=='driver':
                        self.assertNotEqual(entry,verifier.expected_entry(job,root,cell))
                        self.assertTrue(child.returncode!=0 or observed[-1]!=['driver',int(cell[-1]) if cell.startswith('H') else 'N'])
                    else:
                        self.assertNotEqual(child.returncode,0)
                        self.assertEqual(observed,[],child.stderr)
                    if mutation:MUTANTS.append(dict(id=f'{cell}-{mutation}',kind='actual loaded entry stub',detected=True))

    def test_failure_timeout_and_single_use(self):
        materials,bindings=runner.archive_materials()
        for timeout in [False,True]:
            with tempfile.TemporaryDirectory() as tmp:
                output=Path(tmp).resolve()/'run'
                class FakeChild:
                    pid=123
                    returncode=1
                    def wait(self,timeout=None):
                        if timeout is not None and should_timeout:raise subprocess.TimeoutExpired('stub',timeout)
                        return self.returncode
                    def kill(self):self.returncode=-9
                should_timeout=timeout
                fake_command=subprocess.CompletedProcess([],0,'{}','')
                with patch.object(runner,'OUTPUT',output),patch.object(runner,'archive_materials',return_value=(materials,bindings)),patch.object(runner.subprocess,'run',return_value=fake_command),patch.object(runner.subprocess,'Popen',return_value=FakeChild()) as spawn,patch.object(runner.subprocess,'check_output',return_value=b'{"node":"v24.18.0","v8":"13.6.233.17-node.50"}'):
                    self.assertEqual(runner.run(),1)
                    self.assertEqual(spawn.call_count,1)
                    with self.assertRaises(ValueError):runner.run()
                result=runner.read(output/'result.json')
                self.assertEqual(result['status'],'incomplete')
                self.assertEqual(result['notRun'],list(range(1,12)))
                self.assertEqual(runner.read(output/'0-H0'/'exit.json')['exitCode'],-9 if timeout else 1)
                MUTANTS.append(dict(id='timeout' if timeout else 'child-failure',kind='runner child stub',detected=True))

    def test_full_verifier_and_reindexed_negatives(self):
        import shutil
        with tempfile.TemporaryDirectory() as tmp:
            base=Path(tmp).resolve()/'base';base.mkdir();full_fixture(base)
            self.assertEqual(verifier.verify(base,runner.h.ARCHIVE)['samples'],28800)
            for mutation in ['whole-span','thirteenth-process','manifest-function','manifest-indices','manifest-kind','wrong-preflight','missing-completion','summary','environment']:
                root=base.parent/mutation;shutil.copytree(base,root)
                if mutation=='whole-span':
                    for p,c in enumerate(runner.SCHEDULE):
                        f=root/f'{p}-{c}'/'exit.json';r=runner.read(f);r.update(startMonotonic=10+p*100,endMonotonic=11+p*100);write(f,r)
                if mutation=='thirteenth-process':(root/'12-H0').mkdir()
                if mutation.startswith('manifest'):
                    f=root/'derivation.json';r=runner.read(f)
                    if mutation=='manifest-function':r['functionDigests']['runRow']='bad'
                    if mutation=='manifest-indices':r['replacedStatementIndices']=[]
                    if mutation=='manifest-kind':r['kind']='bad'
                    write(f,r)
                if mutation=='wrong-preflight':write(root/'1-N0/preflight-first.json',dict(module=1,pid=123))
                if mutation=='missing-completion':(root/'0-H0/completion.json').unlink()
                if mutation=='summary':
                    f=root/'result.json';r=runner.read(f);r['rows'][0]['candidateOverReference']=1;write(f,r)
                index_fixture(root)
                with patch.dict(os.environ,{'NODE_OPTIONS':'--import=/never-load.mjs'} if mutation=='environment' else {}):
                    with self.assertRaises((ValueError,OSError)):verifier.verify(root,runner.h.ARCHIVE)
                MUTANTS.append(dict(id=mutation,kind='complete reindexed verifier input',detected=True))

if __name__=='__main__':
    result=unittest.TextTestRunner(verbosity=2).run(unittest.defaultTestLoader.loadTestsFromTestCase(NeutralTests))
    receipt=os.environ.get('NEUTRAL_PY_RECEIPT')
    if receipt:
        with Path(receipt).open('x') as f:json.dump(dict(testsRun=result.testsRun,failures=len(result.failures),errors=len(result.errors),mutants=MUTANTS,realConsumerMeasurements=0),f,indent=2)
    raise SystemExit(0 if result.wasSuccessful() else 1)

File: scripts/causal-neutral-driver.test.mjs
/** Loaded private-driver qualification. All factories are stubs; no consumer sampling. */
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import { syncBuiltinESMExports } from "node:module";
import os from "node:os";
import path from "node:path";
import { performance } from "node:perf_hooks";
import timers from "node:timers/promises";
import { fileURLToPath, pathToFileURL } from "node:url";
import { check } from "./check-causal-neutral-source.mjs";
import { derive } from "./derive-causal-neutral-driver.mjs";

const self = fileURLToPath(import.meta.url);
if (process.argv[2] === "--stub") {
	const [driver, cell, mutation] = process.argv.slice(3);
	const { runRow } = await import(pathToFileURL(driver));
	const reference = cell.endsWith("0") ? 0 : 1;
	const order = cell.startsWith("R") ? [1 - reference, reference] : [reference, 1 - reference];
	const config = {
		row: { id: "cold-P2-summary", group: "cold", profile: "P2", mode: "summary" },
		control: true,
		output: "/stub",
		scenarioPath: "scenario",
		copyModule: pathToFileURL(
			path.join(path.dirname(driver), reference === 0 ? "worker-copy.mjs" : "worker.mjs"),
		).href,
		diagnostic: { cell, referenceModule: reference, preflightOrder: order },
	};
	if (mutation === "wrong-row") config.row.group = "recovery";
	if (mutation === "wrong-control") config.control = false;
	if (mutation === "wrong-slot") config.diagnostic.referenceModule = 1 - reference;
	if (mutation === "wrong-cell") config.diagnostic.cell = "H0";
	if (mutation === "copy-self")
		config.copyModule = pathToFileURL(
			path.join(path.dirname(driver), reference === 0 ? "worker.mjs" : "worker-copy.mjs"),
		).href;
	const events = [],
		samples = [],
		outputs = {};
	let now = 0,
		serial = 0;
	const recipe = {
		orders: [
			["candidate", "reference"],
			["reference", "candidate"],
			["candidate", "reference"],
		],
		warmup: 100,
		measured: 300,
	};
	const modules = [0, 1].map((id) => ({
		RECIPE: recipe,
		preflight() {
			events.push(["preflight", id]);
			return { passed: true };
		},
		measurementArm(arm, mode) {
			assert.equal(arm, "reference");
			assert.equal(mode, "summary");
			const instance = ++serial;
			events.push(["factory", id, instance]);
			return {
				cleanup() {
					events.push(["cleanup", instance]);
					if (mutation === "dual-failure") throw Error("cleanup");
				},
			};
		},
	}));
	fs.readFileSync = (p) => JSON.stringify(p === "config" ? config : {});
	fs.writeFileSync = (p, s) => {
		outputs[p] = JSON.parse(s);
	};
	fs.appendFileSync = (_p, s) => {
		events.push(["record"]);
		if (mutation === "dual-failure") throw Error("record");
		samples.push(JSON.parse(s));
	};
	timers.setImmediate = async () => {
		events.push(["tick"]);
	};
	Object.defineProperty(performance, "now", {
		configurable: true,
		value: () => {
			events.push(["clock"]);
			return ++now;
		},
	});
	process.memoryUsage = () => {
		events.push(["memory"]);
		return {};
	};
	syncBuiltinESMExports();
	let failure;
	try {
		await runRow("config", mutation === "same-namespace" ? [modules[0], modules[0]] : modules);
	} catch (e) {
		failure = e;
	}
	const invalid = [
		"wrong-row",
		"wrong-control",
		"wrong-slot",
		"wrong-cell",
		"copy-self",
		"same-namespace",
	].includes(mutation);
	if (invalid) {
		assert.ok(failure);
		assert.deepEqual(events, []);
	} else if (mutation === "dual-failure") {
		assert.ok(failure instanceof AggregateError);
		assert.deepEqual(
			failure.errors.map((e) => e.message),
			["record", "cleanup"],
		);
		assert.equal(serial, 1);
	} else {
		assert.equal(failure, undefined);
		assert.deepEqual(
			events.slice(0, 2),
			order.map((id) => ["preflight", id]),
		);
		const start = events.findIndex((e) => e[0] === "tick");
		assert.equal(events.slice(2, start).filter((e) => e[0] === "preflight").length, 0);
		let cursor = start,
			instance = 0;
		for (const arms of recipe.orders)
			for (const arm of arms)
				for (let i = 0; i < 400; i++) {
					instance++;
					assert.deepEqual(events.slice(cursor, cursor + 8), [
						["tick"],
						["memory"],
						["clock"],
						["factory", arm === "reference" ? reference : 1 - reference, instance],
						["clock"],
						["memory"],
						["record"],
						["cleanup", instance],
					]);
					cursor += 8;
				}
		assert.equal(cursor, events.length);
		assert.equal(samples.length, 2400);
		for (const s of samples) {
			assert.equal(s.ms, 1);
			assert.equal(s.end - s.start, 1);
			assert.equal(s.constructionMs, 0);
			assert.equal(s.preparationMs, 0);
		}
		assert.deepEqual(outputs["/stub/completion.json"], { completed: true, samples: 2400 });
	}
	console.log("STUB_QUALIFIED", cell, mutation);
} else {
	const tmp = fs.realpathSync(fs.mkdtempSync(path.join(os.tmpdir(), "neutral-qual-")));
	const receipt = {
		kind: "loaded-neutral-driver-qualification-v1",
		positiveCells: [],
		staticMutations: [],
		loadedMutations: [],
		realConsumerMeasurements: 0,
	};
	try {
		const extracted = spawnSync(
			"python3",
			[
				"-c",
				'import tarfile,hashlib; t=tarfile.open("archive/evals/causal-performance-repetition-v2/evidence.tar.gz"); print(next(b.decode() for m in t.getmembers() if m.isfile() for b in [t.extractfile(m).read()] if hashlib.sha256(b).hexdigest()=="d9d8606d73e1bcfad66097f027d63c99fc409571042c5f1e7efffaa645356fb2"),end="")',
			],
			{ encoding: "utf8", maxBuffer: 8 * 1024 * 1024 },
		);
		assert.equal(extracted.status, 0, extracted.stderr);
		const source = extracted.stdout;
		const { output } = derive(source);
		assert.equal(check(source, output).valid, true);
		const driver = path.join(tmp, "neutral-driver.mjs");
		const run = (text, cell, mutation) => {
			fs.writeFileSync(driver, text);
			return spawnSync(process.execPath, [self, "--stub", driver, cell, mutation], {
				encoding: "utf8",
				timeout: 15000,
				env: { PATH: process.env.PATH },
			});
		};
		for (const cell of ["N0", "R0", "N1", "R1"]) {
			const r = run(output, cell, "none");
			assert.equal(r.status, 0, r.stderr);
			receipt.positiveCells.push(cell);
		}
		for (const mutation of [
			"wrong-row",
			"wrong-control",
			"wrong-slot",
			"wrong-cell",
			"copy-self",
			"same-namespace",
			"dual-failure",
		]) {
			const r = run(output, "N0", mutation);
			assert.equal(r.status, 0, r.stderr);
			receipt.loadedMutations.push({ mutation, detected: true });
		}
		const replacements = {
			"wrong-factory-slot": [
				'arm === "reference" ? diagnostic.referenceModule',
				'arm === "candidate" ? diagnostic.referenceModule',
			],
			"reversed-preflight": [
				"modules[diagnostic.preflightOrder[0]].preflight",
				"modules[diagnostic.preflightOrder[1]].preflight",
			],
			"missing-preflight": [
				"modules[diagnostic.preflightOrder[1]].preflight(row, scenario)",
				"checked3",
			],
			"missing-cleanup": [
				"if (run !== shared) cleanupAll([run], sampleFailure);",
				"if (false) cleanupAll([run], sampleFailure);",
			],
			"cached-instance": [
				"const makeArm = (arm, mode) => modules[",
				"let cached; const makeArm = (arm, mode) => cached ??= modules[",
			],
			"outside-clock": [
				"const start = performance2.now();",
				"if(row.group === 'cold') run = makeArm(arm,row.mode); const start = performance2.now();",
			],
			generator: ["export async function runRow", "export async function* runRow"],
			"wrong-duration": ["ms: end - start,", "ms: 0,"],
		};
		for (const [mutation, [from, to]] of Object.entries(replacements)) {
			assert.ok(output.includes(from), mutation);
			const altered = output.replace(from, to);
			assert.throws(() => check(source, altered));
			receipt.staticMutations.push({ mutation, detected: true });
			const r = run(altered, "N0", "none");
			assert.notEqual(r.status, 0, mutation);
			receipt.loadedMutations.push({ mutation, detected: true });
		}
		assert.throws(() => check(source, output + "\nconsole.log('extra');"));
		receipt.staticMutations.push({ mutation: "extra-top-level", detected: true });
		console.log(JSON.stringify(receipt, null, 2));
		if (process.env.NEUTRAL_JS_RECEIPT)
			fs.writeFileSync(process.env.NEUTRAL_JS_RECEIPT, JSON.stringify(receipt, null, 2) + "\n", {
				flag: "wx",
			});
	} finally {
		fs.rmSync(tmp, { recursive: true, force: true });
	}
}
